src/main/scala/uk/gov/nationalarchives/DynamoFormatters.scala

```diff
@@ -0,0 +1,154 @@
+    override def read(dynamoValue: DynamoValue): Either[DynamoReadError, DynamoTable] = {
+      val map = dynamoValue.toAttributeValue.m().asScala
+
+      def readNumber[T](name: String, toNumberFunction: String => T): ValidatedNel[InvalidProperty, Option[T]] = {
+        map
+          .get(name)
+          .map { value =>
+            Option(value.n()) match {
+              case Some(value) if Try(toNumberFunction(value)).isSuccess => Option(toNumberFunction(value)).validNel
+              case None => (name -> NoPropertyOfType("Number", DynamoValue.fromAttributeValue(value))).invalidNel
+              case Some(value) =>
+                (name -> TypeCoercionError(
+                  new RuntimeException(s"Cannot parse $value for field $name to a number")
+                )).invalidNel
+            }
+          }
+          .getOrElse(None.validNel)
+      }
+      def valueOrInvalid(name: String): ValidatedNel[InvalidProperty, String] = value(name)
+        .map(_.validNel)
+        .getOrElse((name -> MissingProperty).invalidNel)
+
+      def value(name: String) = map.get(name).map(_.s())
+
+      (
+        valueOrInvalid(batchId),
+        valueOrInvalid(id),
+        valueOrInvalid(name),
+        stringToType(value(typeField)),
+        readNumber(fileSize, _.toLong),
+        readNumber(sortOrder, _.toInt)
+      ).mapN { (batchId, id, name, typeName, fileSize, sortOrder) =>
+        val identifiers = map
+          .filter(_._1.startsWith("id_"))
+          .map { case (name, value) =>
+            Identifier(name.drop(3), value.s())
+          }
+          .toList
+        DynamoTable(
+          batchId,
+          UUID.fromString(id),
+          value(parentPath),
+          name,
+          typeName,
+          value(title),
+          value(description),
+          sortOrder,
+          fileSize,
+          value(checksumSha256),
+          value(fileExtension),
+          identifiers
+        )
+      }.toEither
+        .left
+        .map(InvalidPropertiesError.apply)
+    }
+
+    override def write(t: DynamoTable): DynamoValue = {
+      val potentialValues = Map(
+        batchId -> Option(DynamoValue.fromString(t.batchId)),
+        id -> Option(DynamoValue.fromString(t.id.toString)),
+        parentPath -> t.parentPath.map(DynamoValue.fromString),
+        name -> Option(DynamoValue.fromString(t.name)),
+        typeField -> Option(DynamoValue.fromString(t.`type`.toString)),
+        title -> t.title.map(DynamoValue.fromString),
+        description -> t.description.map(DynamoValue.fromString),
+        sortOrder -> t.sortOrder.map(DynamoValue.fromNumber[Int]),
+        fileSize -> t.fileSize.map(DynamoValue.fromNumber[Long]),
+        checksumSha256 -> t.checksumSha256.map(DynamoValue.fromString),
+        fileExtension -> t.fileExtension.map(DynamoValue.fromString)
+      ) ++ t.identifiers.map(id => s"id_${id.identifierName}" -> Option(DynamoValue.fromString(id.value))).toMap
+      val values = for {
```

Comment:
I've just tried to compile and it's complaining that the match is not exhausive and I don't want to turn off the warnings as errors thing.

We can have:
```scala
      val values = potentialValues.flatMap {
        case (fieldName, Some(potentialValue)) => Map(fieldName -> potentialValue)
        case _ => Map.empty
      }
```
I think that's clearer than the for comprehension. What do you think?

Reply:
Ah yeah, that makes sense, I'm guessing it's considered non-exhaustive because, there isn't a case for `None`? but as you said, that shouldn't happen? Maybe we should rename `potentialValues ` and/or `potentialValue`?